### ISSUE TITLE
Device configuration for protocols without a discovery mechanism

### ISSUE THREAD

**Issue by sogaani** [mozilla-iot/gateway#993]
If devices which does not have discovering mechanism, e.g. KNX, Gpio pin, and Virtual devices, we should configure to decide the thing descriptions.

### Now
We have configuration API for the add-on, and an add-on can provide the configuration page for their devices.

#### add device sequence
1. The user go the add-on configuration page and configure device.
2. The user go the add thing page, find device and save device.

#### remove device sequence
1. The user go the things page, right click on device icon, and press remove.
2. The user go the add-on configuration page and delete configuration for device.

### I want
We have device configuration API for the add-on, and an add-on can provide input form like thing-url add-on on add thing page.

#### add device sequence
2. The user go the add thing page, input forms and save device.

#### remove device sequence
1. The user go the things page, right click on device icon, and press remove.

@benfrancis wrote in https://github.com/mozilla-iot/gateway/issues/668
> The kind of flow you're describing where arbitrary user input is described using JSON Schema sounds a lot like an action input on the Web Thing API. Perhaps that could be implemented as an action on the gateway, where the gateway gets its own Thing Description (something we talked about before). 

Good advice, I agree that we can implement above sequence with actions. Since, the input forms for device configuration can be different per add-ons, probably an action input is formed the below or actions per add-ons is needed.

```
"configDevice": {
  "description": "Configure devices",
  "input": {
    "type": "object",
    "properties": {
      "GPIO": {
        "type": "object",
        "properties": {
          ....
        }
      },
      "KNX": {
        "type": "object",
        "properties": {
           ...
        }
      }
    }
  },
  "href": "/gateway/actions/configDevice"
}
```

**Comment by benfrancis** [comment]
Thank you for explaining these use cases in a bit more detail.

Actions could be one way of implementing this at the API level, another way could be to modify the `POST /things` API to allow additional information to be submitted when adding certain specialised types of thing.

From a UI point of view this sounds similar to manually adding a web thing by its URL, but the UI would need augmenting to provide different forms for different specialised types of thing, which can be extended by add-ons. One approach would be to offer multiple options when clicking the "+" button in the things screen, to manually add a thing by its URL, manually add a GPIO thing, manually add a KNX thing or to scan for things using adapters.

I think this requires some further exploration.

**Comment by sogaani** [comment]
@benfrancis 
I made [UI mock](https://docs.google.com/presentation/d/1ykD_xxDT39H4Ja67ANp01vQRUo8vQM_9D7YAZl5y6zM/edit?usp=sharing). How do you feel about this flow and UI?

**Comment by benfrancis** [comment]
@sogaani wrote:
> I made UI mock. How do you feel about this flow and UI?

Thank you for this.

A couple of things I don't like:
a) It clutters the Add Device screen which already has an "Add by URL..." link
b) It exposes a lot of device specific technical details to the web layer

To address problem A, a slightly different approach for the UI might be to have an "Add device manually..." link instead of an "Add by URL..." link which then adds an empty entry to the list which accepts a URL by default but can be changed (e.g. via a drop down menu) to add a device using another mechanism provided by an adapter.

With regards to problem B, in general we want to abstract away implementation details like pin numbers from the web layer, but I accept that in some cases where devices need to be added manually rather than automatically discovered by an adapter, this will be challenging.

Another question is how an adapter add-on can expose the configuration parameters it requires for a manually added device, though I suspect you're proposing somehow using JSON Schema for that, from which a basic form can be generated automatically?

**Comment by benfrancis** [comment]
@mishoboss wrote:
> Also, I would like to give an idea. You know, there is a migration now to a more flexible capability system. 

This does complicate things a little. Usually an adapter add-on would be responsible for automatically detecting a device and classifying it to generate a Thing Description which describes the capabilities that thing exposes in Web of Things terms.

In the case of manually adding a device, the user somehow needs to map capabilities onto a device and the method for doing that is likely to be very device specific.

**Comment by mishoboss** [comment]
> In the case of manually adding a device, the user somehow needs to map capabilities onto a device and the method for doing that is likely to be very device specific.

Sure, it has to be device specific. In KNX we describe Group Addresses, in GPIO we describe pin numbers and IN/OUT directions, in MQTT we describe channels and value parsing regex, etc. 
This is the reason we need per-device JSON-SCHEMA describing the necessary input fields. But before UI populates those fields, on the previous step, the user must select what capabilities will be described. That's my point. Because the user may describe a complex device like a thermostat or roller shutter, where we have several properties. Without that step, I can't imagine how the hub will map those custom described devices to something meaningful.

**Comment by sogaani** [comment]
@mishoboss wrote:
> Why? IMHO, let's keep that page clean. Just for adding things, not listing already added things. We have a dedicated page for listing added things, right?

This is good for user. However we need name displayed for adding thing. Furthermore, capability system has to [choose device type](https://github.com/mozilla-iot/gateway/issues/1090). When the Gateway is adding thing, can not know what type the device have. And this is same UI process as "Add by URL...". I think that same process give user good UI experience.


@benfrancis  wrote:
> To address problem A, a slightly different approach for the UI might be to have an "Add device manually..." link instead of an "Add by URL..." link which then adds an empty entry to the list which accepts a URL by default but can be changed (e.g. via a drop down menu) to add a device using another mechanism provided by an adapter.

I like this. As I said the above, I think same UI process as existing give user better UI experience. Updata [UI MockUp](https://docs.google.com/presentation/d/1P-ltJYy6YQH6fU9o2-XW7JJMKb5JZ_dIGmev_0KBxjA/edit#slide=id.g39f1f7875a_0_37).

> With regards to problem B, in general we want to abstract away implementation details like pin numbers from the web layer, but I accept that in some cases where devices need to be added manually rather than automatically discovered by an adapter, this will be challenging.

Is this care about security or UI experience? I'm optimistic about UI experience. Because, I think that the user using iot device which not have automatically discovered mechanism well know about devices.

> Another question is how an adapter add-on can expose the configuration parameters it requires for a manually added device, though I suspect you're proposing somehow using JSON Schema for that, from which a basic form can be generated automatically?

My thinking, An adapter can expose the configuration parameters by JSON schema like as [the addon configuration](https://github.com/mozilla-iot/wiki/wiki/Add-On-Configuration). An adapter has asDict method and can return JSON schema via asDict. And the Gateway can expose it as action input.

Example
```
{
  "id": "gpioAdapterId",
  "name": "gpioAdapter",
  "ready": ture,
  "manualDevice": {
    "type": "object",
    "required": [
      "pin"
    ],
    "properties": {
      "pin": {
        "type": "integer",
        "minimum": 0
      },
      "name": {
        "type": "string"
      },
      "direction": {
        "type": "string",
          "enum": [
            "in",
            "out"
          ]
        },
        "value": {
          "type": "number",
          "enum": [
            0,
            1
          ]
        },
        "edge": {
          "type": "string",
          "enum": [
            "none",
            "rising",
            "falling",
            "both"
          ]
        },
        "debounce": {
          "type": "number",
          "minimum": 0
        }
      }
    }
};
```

@mishoboss wrote:
> Also, I would like to give an idea. You know, there is a migration now to a more flexible capability system.

IMHO, the addon only support major type devices (e.g. shutter, thermostat, light on KNX) rather than capability. The capability is machine readable but some capability is hard to understand for users. Using JSON schema with [`oneof` option](https://github.com/mozilla-iot/wiki/wiki/Add-On-Configuration#configuration-which-have-dynamic-form) expose configuration per device types.

**Comment by mishoboss** [comment]
> IMHO, the addon only support major type devices (e.g. shutter, thermostat, light on KNX) rather than capability. The capability is machine readable but some capability is hard to understand for users. Using JSON schema with oneof option expose configuration per device types.

Well, after all, device types are just containers that define some set of capabilities. So if we do not "oneof", but multi-select of device types, we can enrich the set of capabilities that can be manually defined. If I correctly understand how the idea of the capability system implementation.

BTW, I see something in the updated UI mockup. On slide 6 (the KNX Adapter one) there is a "Device Type" dropdown, but on slide 4 (the GPIO Adapter), there isn't. Why is that? What semantical meaning has some pin number? We need the "Device Type" dropdown for every adapter, not just KNX. Imagine that we can manually create an "onOffLight" thing using the GPIO adapter, and define a pin for it. On that pin we can have a 220V AC relay where a light bulb is hooked. So, semantically this is a "onOffLight" thing with its properties, actions and events. And it can be rendered on UI and participate in rules as any other "onOffLight" light, no matter what protocol/technology is underneath.

**Comment by sogaani** [comment]
> BTW, I see something in the updated UI mockup. On slide 6 (the KNX Adapter one) there is a "Device Type" dropdown, but on slide 4 (the GPIO Adapter), there isn't. Why is that?

Because the GPIO adapter supports only device Type onOffSwitch and binarySensor and it's device type is decided by `direction`.
Supporting device or property capabilities are depend and restricted on the adapter. If trying supports color property on the GPIO adapter, there is many implementations.

> So, semantically this is a "onOffLight" thing with its properties, actions and events.

It is good idea. However, the GPIO Adapter should implemented how it work. Lights may have Off circuit or Self-holding circuit. Furthermore, number property or string property have more complexly.

I think that If the Adapter supports some selectable property or device type, the Adapter express these capability by JSON-schema or some mechanism.

**Comment by mishoboss** [comment]
> Because the GPIO adapter supports only device Type onOffSwitch and binarySensor and it's device type is decided by direction.

Well, not necessary. In reality it supports anything binary. That could be `onOffSwitch`, `binarySensor` and `onOffLight` for now. Maybe some more in future. The point is to provide a semantical meaning of that pin. And render a unified UI approach for setting that semantical meaning. After user choses the capability, a JSON-SCHEMA form will be rendered with necessary fields to fill capability with the technology related stuff - pin, KNX GAs, MQTT topics, whatever the tech relies on.

**Comment by mishoboss** [comment]
BTW, taking the above approach on a next level will mean that GPIO could not only support binary stuff, but even more complex capabilities. Imagine a "Roller Shutter" capability. It needs a pin for UP movement and a pin for DOWN movement (imagine someone hooked there two relays and drives a shutter motor). So we can build now more a complex device out of just GPIO adapter. And same thing applies for the MQTT adapter, just the user doesn't supply UP pin and DOWN pin, but supplies UP MQTT topic and DOWN MQTT topic.

**Comment by benfrancis** [comment]
@sogaani wrote:
> I think same UI process as existing give user better UI experience. Updata UI MockUp.

I think this is a good direction. I would make adding a native web thing by its URL the default option in this list. Maybe even hide the drop-down box if no other add-ons are installed which support manual device creation.

I am slightly concerned about how complex this UI could become. A UI for manually adding capabilities, properties, actions and events to Things and mapping those onto functions of different underlying protocols is almost a developer IDE for creating web things. This is probably too complex for most end users.

**Comment by mishoboss** [comment]
> I am slightly concerned about how complex this UI could become. A UI for manually adding capabilities, properties, actions and events to Things and mapping those onto functions of different underlying protocols is almost a developer IDE for creating web things. This is probably too complex for most end users.

Well, giving the fact that most other HA/IoT software out there don't even provide a GUI for this kind of stuff, but instead they require some YAML/XML/JSON description in a file. So a GUI is a huge step forward and would be considered way less geeky than almost any competition known to the moment! :) But its the only way to integrate some technologies. 
Yes, this is not something a regular user would do. KNX is not off-the-shelf product you can buy from Wallmart and just plug it. MQTT devices are usually DIY. GPIO too. So dealing with those kind of devices usually means you know what you do. In Mozilla IoT project I see a potential that to be handled the right way this time.

**Comment by benfrancis** [comment]
Something I think makes Firefox very powerful is that it has a very simple and clean default UI that anyone can use, but it can be heavily customised and extended by power users through add-ons.

I would like to enable the same level of usability and flexibility in the Things Gateway. So as long as it doesn't compromise the simplicity of the default UI, I think it's fine to add additional complexity through add-ons. We just need to make sure the default experience is not overcomplicated, e.g. by not showing the drop-down box for different types of thing to manually create unless the user has installed add-ons which add new types of device that can be manually added.

**Comment by sogaani** [comment]
> I think this is a good direction. I would make adding a native web thing by its URL the default option in this list. Maybe even hide the drop-down box if no other add-ons are installed which support manual device creation.

I agree. We should not display the link adding manual devices, If there is no addon providing manual devices.

> That said, the minimal approach you talk about could be achieved, and add-ons are the ones that embed the whole logic necessary for the manual configuration of the devices they support. UI just renders the properties by the selected capability in a way that is described in add-on specific JSON-SCHEMA.

We have some capabilities "devices, properties, and actions". Your suggestion seems to suppose only properties. In this case, my concern is who decided the device type and how exposing the capabilities.

If an addon decide the device type, the addon could  decide with properties and user inputs or just support `thing`. If the Gateway decide the device type, I guess sometime deciding device type correctly is difficult. For example the GPIO adapter only supports `on` property and supports "onOffSwitch", "onOffSensor". The GPIO adapter decide device type from user input "`direction`, `edge`".

If we display supported capabilities in unified way through the addons, the addons need to expose capabilities. I think exposing capability list with JSON is natural. @mishoboss do you have any suggestion for this?

JSON capability list sample with GPIO adapter.
```
{
  "capabilities": {
    "devices": [ // device capabilities
      "onOffLight",
      "onOffSwitch",
      "onOffSensor"
    ],
    "properties": { // property capabilities
      "on": { // JSON-schema for on property
        "pin": {
          "type": "integer",
          "minimum": 0
        },
        "direction": {
          "type": "string",
          "enum": [
            "in",
            "out"
          ]
        },
        "value": {
          "type": "number",
          "enum": [
            0,
            1
          ]
        },
        "edge": {
          "type": "string",
          "enum": [
            "none",
            "rising",
            "falling",
            "both"
          ]
        },
        "debounce": {
          "type": "number",
          "minimum": 0
        }
      }
    },
    "actions": {}
  }
}
```

**Comment by mishoboss** [comment]
Hi, if I get your questions correctly, I think that the addon is the one that provides what device types (capabilities) are available for user to choose from. Then these are rendered in a list (possibly a dropdown or even a multi-selection dropdown) and user is the one that selects the device type (capabilities) that he wants to add.

I checked your example, but I can't understand why you supply properties and actions there. They shouldn't be defined there. They are part of the capabilities and we have defined a capabilities list with actions, events and properties. The addon just needs to specify which of them it can support. So we are only interested in what kind of capabilities the addon can deal with. I imagine something like that:

```
{
    "devices": [ // device capabilities
      "onOffLight",
      "onOffSwitch",
      "onOffSensor"
    ],
    "fields": { // config fields per property
       ... JSON-SCHEMA
    }     
}
```
In the JSON-SCHEMA we define addon specific input fields that are necessary to config each property from the chosen capability. For example lets assume that a user selects "onOffLight", so we have just "on" property. The JSON-SCHEMA defines two input fields per each available property in a capability. In this case we have just one property, so UI renders just two fields - "PIN" and "DIRECTION". Those fields are defined in the JSON-SCHEMA that is provided by the addon. Each addon knows what fields it needs to supply, because the addon is the one that understands its underneath technology specifics. For example, for a MQTT addon, the JSON-SCHEMA will define "topic" input field and "regex value" input field per each property that a capability implements. For KNX addon, there would be "Read GA" and "Write GA" for each property. And so on. :)
